File: health-checker/config/config_test.go
package config

import (
	"testing"
)

func TestLoadDefaults(t *testing.T) {
	// Clear all env vars the loader consults so we exercise the defaults.
	for _, k := range []string{
		"CHECKER_PORT",
		"LOG_LEVEL",
		"API_GATEWAY_URL",
		"ALERT_SERVICE_URL",
		"CHECK_TIMEOUT",
		"READ_HEADER_TIMEOUT",
		"READ_TIMEOUT",
		"WRITE_TIMEOUT",
		"IDLE_TIMEOUT",
		"SHUTDOWN_TIMEOUT",
	} {
		t.Setenv(k, "")
	}

	cfg := Load()

	cases := map[string]string{
		"Port":              cfg.Port,
		"LogLevel":          cfg.LogLevel,
		"APIGatewayURL":     cfg.APIGatewayURL,
		"AlertServiceURL":   cfg.AlertServiceURL,
		"CheckTimeout":      cfg.CheckTimeout,
		"ReadHeaderTimeout": cfg.ReadHeaderTimeout,
		"ReadTimeout":       cfg.ReadTimeout,
		"WriteTimeout":      cfg.WriteTimeout,
		"IdleTimeout":       cfg.IdleTimeout,
		"ShutdownTimeout":   cfg.ShutdownTimeout,
	}
	want := map[string]string{
		"Port":              "8001",
		"LogLevel":          "INFO",
		"APIGatewayURL":     "http://api-gateway:8000",
		"AlertServiceURL":   "http://alert-service:8002",
		"CheckTimeout":      "5s",
		"ReadHeaderTimeout": "5s",
		"ReadTimeout":       "15s",
		"WriteTimeout":      "15s",
		"IdleTimeout":       "60s",
		"ShutdownTimeout":   "30s",
	}

	for field, got := range cases {
		if got != want[field] {
			t.Errorf("Load() %s = %q, want %q", field, got, want[field])
		}
	}
}

func TestLoadOverridesFromEnv(t *testing.T) {
	overrides := map[string]string{
		"CHECKER_PORT":        "9001",
		"LOG_LEVEL":           "DEBUG",
		"API_GATEWAY_URL":     "http://gw:1",
		"ALERT_SERVICE_URL":   "http://alerts:2",
		"CHECK_TIMEOUT":       "1s",
		"READ_HEADER_TIMEOUT": "2s",
		"READ_TIMEOUT":        "3s",
		"WRITE_TIMEOUT":       "4s",
		"IDLE_TIMEOUT":        "5s",
		"SHUTDOWN_TIMEOUT":    "6s",
	}
	for k, v := range overrides {
		t.Setenv(k, v)
	}

	cfg := Load()

	if cfg.Port != "9001" {
		t.Errorf("Port = %q, want 9001", cfg.Port)
	}
	if cfg.LogLevel != "DEBUG" {
		t.Errorf("LogLevel = %q, want DEBUG", cfg.LogLevel)
	}
	if cfg.APIGatewayURL != "http://gw:1" {
		t.Errorf("APIGatewayURL = %q", cfg.APIGatewayURL)
	}
	if cfg.AlertServiceURL != "http://alerts:2" {
		t.Errorf("AlertServiceURL = %q", cfg.AlertServiceURL)
	}
	if cfg.CheckTimeout != "1s" {
		t.Errorf("CheckTimeout = %q", cfg.CheckTimeout)
	}
	if cfg.ReadHeaderTimeout != "2s" {
		t.Errorf("ReadHeaderTimeout = %q", cfg.ReadHeaderTimeout)
	}
	if cfg.ReadTimeout != "3s" {
		t.Errorf("ReadTimeout = %q", cfg.ReadTimeout)
	}
	if cfg.WriteTimeout != "4s" {
		t.Errorf("WriteTimeout = %q", cfg.WriteTimeout)
	}
	if cfg.IdleTimeout != "5s" {
		t.Errorf("IdleTimeout = %q", cfg.IdleTimeout)
	}
	if cfg.ShutdownTimeout != "6s" {
		t.Errorf("ShutdownTimeout = %q", cfg.ShutdownTimeout)
	}
}

File: alert-service/src/config.test.ts
import { loadConfig, resolveLogLevel, resolvePort } from './config';

const ENV_KEYS = ['ALERT_PORT', 'LOG_LEVEL', 'API_GATEWAY_URL', 'ALERT_DEDUP_WINDOW_SECONDS'];

/**
 * config は環境変数を都度読むため、各テストの前後で対象キーを退避・復元する。
 */
function withEnv<T>(overrides: Record<string, string | undefined>, fn: () => T): T {
  const saved: Record<string, string | undefined> = {};
  for (const key of ENV_KEYS) saved[key] = process.env[key];
  try {
    for (const [k, v] of Object.entries(overrides)) {
      if (v === undefined) delete process.env[k];
      else process.env[k] = v;
    }
    // Clear any keys not explicitly overridden so we exercise defaults.
    for (const key of ENV_KEYS) {
      if (!(key in overrides)) delete process.env[key];
    }
    return fn();
  } finally {
    for (const key of ENV_KEYS) {
      if (saved[key] === undefined) delete process.env[key];
      else process.env[key] = saved[key]!;
    }
  }
}

describe('resolvePort', () => {
  it.each([
    ['1', 1],
    ['8002', 8002],
    ['9001', 9001],
    ['65535', 65535],
  ])('accepts valid port %s', (raw, expected) => {
    expect(resolvePort(raw)).toBe(expected);
  });

  it.each([undefined, ''])('falls back to default when unset (%s)', (raw) => {
    expect(resolvePort(raw)).toBe(8002);
  });

  it.each(['abc', '8002abc', '8.5', '0x1F', '-'])(
    'falls back for non-integer %s and warns',
    (raw) => {
      const spy = jest.spyOn(process.stderr, 'write').mockImplementation(() => true);
      try {
        expect(resolvePort(raw)).toBe(8002);
        expect(spy).toHaveBeenCalled();
        const msg = spy.mock.calls.map((c) => String(c[0])).join('');
        expect(msg).toContain('invalid ALERT_PORT');
      } finally {
        spy.mockRestore();
      }
    },
  );

  it.each(['0', '-1', '65536', '999999'])(
    'falls back for out-of-range %s and warns',
    (raw) => {
      const spy = jest.spyOn(process.stderr, 'write').mockImplementation(() => true);
      try {
        expect(resolvePort(raw)).toBe(8002);
        expect(spy).toHaveBeenCalled();
        const msg = spy.mock.calls.map((c) => String(c[0])).join('');
        expect(msg).toContain('out of range');
      } finally {
        spy.mockRestore();
      }
    },
  );
});

describe('resolveLogLevel', () => {
  it.each([
    ['DEBUG', 'DEBUG'],
    ['INFO', 'INFO'],
    ['WARN', 'WARN'],
    ['ERROR', 'ERROR'],
    ['debug', 'DEBUG'],
    ['  info  ', 'INFO'],
    ['Error', 'ERROR'],
  ])('normalizes %s to %s', (raw, expected) => {
    expect(resolveLogLevel(raw)).toBe(expected);
  });

  it('aliases WARNING to WARN so ops can share LOG_LEVEL with api-gateway', () => {
    expect(resolveLogLevel('WARNING')).toBe('WARN');
    expect(resolveLogLevel('warning')).toBe('WARN');
  });

  it.each([undefined, ''])('falls back to INFO when unset (%s)', (raw) => {
    expect(resolveLogLevel(raw)).toBe('INFO');
  });

  it.each(['TRACE', 'verbose', 'SHOUT', '1', '!!'])(
    'falls back to INFO for invalid %s and warns',
    (raw) => {
      const spy = jest.spyOn(process.stderr, 'write').mockImplementation(() => true);
      try {
        expect(resolveLogLevel(raw)).toBe('INFO');
        expect(spy).toHaveBeenCalled();
        const msg = spy.mock.calls.map((c) => String(c[0])).join('');
        expect(msg).toContain('invalid LOG_LEVEL');
      } finally {
        spy.mockRestore();
      }
    },
  );
});

describe('loadConfig', () => {
  it('returns defaults when env vars are absent', () => {
    withEnv({}, () => {
      const cfg = loadConfig();
      expect(cfg.port).toBe(8002);
      expect(cfg.logLevel).toBe('INFO');
      expect(cfg.apiGatewayUrl).toBe('http://api-gateway:8000');
      expect(cfg.alertDedupWindowSeconds).toBe(300);
    });
  });

  it('accepts valid overrides', () => {
    withEnv(
      {
        ALERT_PORT: '9002',
        LOG_LEVEL: 'debug',
        API_GATEWAY_URL: 'http://gw:1',
        ALERT_DEDUP_WINDOW_SECONDS: '60',
      },
      () => {
        const cfg = loadConfig();
        expect(cfg.port).toBe(9002);
        expect(cfg.logLevel).toBe('DEBUG');
        expect(cfg.apiGatewayUrl).toBe('http://gw:1');
        expect(cfg.alertDedupWindowSeconds).toBe(60);
      },
    );
  });

  it('does not throw and yields usable defaults when env is malformed', () => {
    // Regression guard: previously ALERT_PORT=abc silently produced NaN and
    // LOG_LEVEL=verbose silently disabled the logger.
    const spy = jest.spyOn(process.stderr, 'write').mockImplementation(() => true);
    try {
      withEnv({ ALERT_PORT: 'not-a-port', LOG_LEVEL: 'verbose' }, () => {
        const cfg = loadConfig();
        expect(cfg.port).toBe(8002);
        expect(cfg.logLevel).toBe('INFO');
      });
    } finally {
      spy.mockRestore();
    }
  });
});
